text/0432-contextual-helpers.md

```diff
@@ -0,0 +1,811 @@
+object can then be passed around like any other values and invoked at a later
+time.
+
+(Typically, a number of them are passed to the `hash` helper to "bundle" them
+into a single object, but this is not required.)
+
+While this is indeed a pretty advanced feature, the users of `SuperForm` do not
+need to be aware of these implementation details in order to use the addon.
+This had proved to be a very useful and powerful feature and enabled a number
+of popular addons, such as
+[ember-cli-addon-docs](https://ember-learn.github.io/ember-cli-addon-docs/),
+[ember-bootstrap](https://www.ember-bootstrap.com),
+[ember-table](https://opensource.addepar.com/ember-table/),
+[ember-paper](https://miguelcobain.github.io/ember-paper/),
+[ember-power-calendar](https://ember-power-calendar.com),
+[ember-accordion](http://khorus.github.io/ember-accordion/),
+[emberx-select](https://emberx-select.netlify.com/),
+[ember-light-table](https://offirgolan.github.io/ember-light-table/).
+
+The original RFC left an important question unanswered – [should this feature
+be available for helpers, too?](https://github.com/emberjs/rfcs/blob/master/text/0064-contextual-component-lookup.md#unresolved-questions)
+
+In this RFC, we argue – yes, this feature will be just as useful for helpers as
+well as modifiers.
+
+For example, the `SuperForm` addon API can be expanded to include some extra
+helpers and modifiers, like so:
+
+```hbs
+<SuperForm @model={{this.post}} as |f|>
+  <f.Input @name="title" />
+
+  {{!-- f.is-valid and f.error-for are contextual helpers --}}
+  {{#unless (f.is-valid "title")}}
+    <div class="error">This field {{f.error-for "title"}}</div>
+  {{/unless}}
+
+  {{!-- f.auto-resize is a contextual modifier --}}
+  <f.Textarea @name="body" {{f.auto-resize maxHeight="500"}} />
+
+  <f.Submit />
+</SuperForm>
+```
+
+For reference, the `<SuperForm>` component's template would look something like
+this:
+
+```hbs
+<form ...>
+  {{yield (hash
+    is-valid=(helper "super-is-valid" form=this model=this.model)
+    error-for=(helper "super-error-for" form=this model=this.model)
+    auto-resize=(modifier "super-auto-resize")
+    ...
+  )}}
+</form>
+```
+
+This RFC proposes a complete design for enabling this capability.
+
+## Detailed design
+
+### The `helper` and `modifier` helpers
+
+This RFC introduces two new helpers named `helper` and `modifier`, which work
+similarly to the `component` helper:
+
+* When passed a string (e.g. `(helper "foo")`) as the first argument, it will
+  produce an opaque, internal "helper definition" or "modifier definition"
+  object that can be passed around and invoked elsewhere.
+
+* Any additional positional and/or named arguments will be stored ("curried")
+  inside the definition object, such that, when invoked, these arguments will
+  be passed along to the referenced helper or modifier.
+
+Some additional details:
+
+* When the first argument passed to the `helper` or `modifier` helper is
```

Comment:
This matches the behavior of the `component` helper. In general, Handlebars tries to be quite forgiving about these things, and errors during render is generally unexpected (and currently unrecoverable).

It also enables use cases like:

```hbs
{{#if this.featureIsEnabled}}
  {{yield (component "experimental")}}
{{else}}
  {{yield null}}
{{/if}}
```

..which the other side can further curry or "invoke" it without caring for whether the value is falsy.

I suppose we can statically disallow the `{{helper ""}}` but I don't quite see the point of it since 1) we have to support the runtime empty string anyway (see above) and 2) this involves using Magic™ that is otherwise not possible in user land (the compiler needs to know that the `helper` helper is special and disallow the static empty string).

I don't think it is worth the trouble.